### PR TITLE
updated fluent

### PR DESCRIPTION
**Pull request by Prince2k3**
Added more clean methods to help build queries, Updated the Model protocol to have more convenient methods. Added StatementGenerator Protocol so that Fluent can have NoSql support later. Adds support for the drivers to handle sql injections by outputting parameterized query and query values as separate properties (That why the drivers can bind the values to the query.)

### CODE REVIEW

File: Sources/Driver.swift [hunk 1]
```diff
@@ -1,10 +1,5 @@
...
 public protocol Driver {
-	func fetchOne(table table: String, filters: [Filter]) -> [String: String]?
-	func fetch(table table: String, filters: [Filter]) -> [[String: String]]
-	func delete(table table: String, filters: [Filter])
-	func update(table table: String, filters: [Filter], data: [String: String])
-	func insert(table table: String, items: [[String: String]])
-	func upsert(table table: String, items: [[String: String]])
-	func exists(table table: String, filters: [Filter]) -> Bool
-	func count(table table: String, filters: [Filter]) -> Int
+    var statementClass: StatementGenerator.Type { get }
```
Comment (ketzusaka) [inline Sources/Driver.swift]: I think this is better served as an "associatedtype" declaration

File: Sources/Model.swift [hunk 2]
```diff
@@ -1,38 +1,101 @@
...
-		in the database.
-	*/
-	func serialize() -> [String: String]
-
-	init(serialized: [String: String])
+    static var entity: String { get }
+    var id: String? { get }
+    
+    func serialize() -> [String: StatementValueType]
+    init(deserialize: [String: String])
```
Comment (ketzusaka) [inline Sources/Model.swift]: Can we init with a `[String: StatementValueType]` as well?

File: Sources/Query.swift [hunk 3]
```diff
@@ -1,121 +1,200 @@
...
+        return self
+    }
+
+    // MARK: - Aggregate
+
+    public func count(key: String = "*") -> Int {
+        guard let result = aggregate(.COUNT(key)) else {
+            return -1
+        }
+        return Int(result["COUNT(\(key))"]!) ?? -1
```
Comment (ketzusaka) [inline Sources/Query.swift]: Individual drivers may call this something else. Is there like a .first?

I'd probably also prefer a `throws` or optional in all these calls vs a magic number

File: Sources/SQL.swift [hunk 4]
```diff
@@ -1,127 +1,287 @@
...
+public class SQL: StatementGenerator {
+    private var tokens: [String] = []
+    private var indexes: [Int] = []
+    private var values: [StatementValueType] = []
+    
+    public var fields: [String]?
+    public var entity: String
+    public var clause: Clause = .SELECT
+    public var operation: [(String, Operator, [StatementValueType])]? = []
+    public var andIndexes: [Int]? = []
```
Comment (ketzusaka) [inline Sources/SQL.swift]: Could these (and all other vars with default values here) be non-optional?

**Comment by Prince2k3** [comment]
@ketzusaka  Thanks! All very good points. I'll evaluate and update. Note:  If you also like to change them I have no problems with that. Being as we are busy people. Still though I like these feedback as it helps me become a better swift developer.

File: Sources/Driver.swift (hunk 1 above)
Comment (Prince2k3) [inline Sources/Driver.swift]: Not sure what you mean by using associatedtype here.

File: Sources/Model.swift [hunk 5]
```diff
@@ -1,38 +1,101 @@
...
+    public static func find(ids: StatementValueType...) -> [Self]? {
+        return Query()._with("id", .In, ids).all()
+    }
+    
+    public static func findBy(key: String, _ op: Operator = .Equals, _ values: StatementValueType...) -> [Self]? {
+        return Query()._with(key, .Equals, values).all()
+    }
+    
+    public static func take(count: Int = 1) -> [Self]? {
+        return Query().limit(1).all()
```
Comment (timominous) [inline Sources/Model.swift]: shouldn't this be `limit(count)`?

File: Sources/Query.swift [hunk 6]
```diff
@@ -1,121 +1,200 @@
...
+public class Query<T: Model> {
+    private(set) var entity: String
+    private var statement: StatementGenerator
+    
+    public init(entity: String? = nil) {
+        self.entity = entity ?? T.entity
+        statement = Database.driver.statementClass.init(entity: self.entity)
+    }
+    
+    public var toSQL: String {
```
Comment (timominous) [inline Sources/Query.swift]: why is there a `toSQL` function in `Query`?

File: Sources/Fluent/Model.swift [hunk 7]
```diff
@@ -0,0 +1,101 @@
+
+public protocol Model {
+    static var entity: String { get }
+    var id: String? { get }
+    
+    func serialize() -> [String: StatementValueType]
```
Comment (tannernelson) [inline Sources/Fluent/Model.swift]: Could we call `StatementValueType` just `Value`? `Type` is unnecessary as of new Swift API Guidelines, statement also seems redundant.

Comment (tannernelson) [inline Sources/Fluent/Model.swift]: Also can you pass `nil` to serialize?

File: Sources/FluentDev/main.swift [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+//: Query Builder - Different looks of using Query Builder Class
+
+// Insert
+
+Query<User>().insert(u.serialize()).run()
+
+// Update
+
+Query<User>().update(u.serialize()).run()
```
Comment (tannernelson) [inline Sources/FluentDev/main.swift]: When you call `.update()` or `.insert()` it should run it. 

Like `Query<User>().with("name", .Equals, "tanner").update("name", "new name")`

**Comment by tannernelson** [comment]
I'm also wondering what the reasoning is for `SQLTokenizer`?

File: Sources/Fluent/Model.swift (hunk 7 above)
Comment (Prince2k3) [inline Sources/Fluent/Model.swift]: Currently the only way to set nil is string NULL. I did the optional route and it confuses the StatementValue and wraps it in an unwrap-able optional.

File: Sources/FluentDev/main.swift (hunk 8 above)
Comment (Prince2k3) [inline Sources/FluentDev/main.swift]: should this be a dictionary instead? or should I have two version of insert or update to allow for a dictionary entry?

**Comment by Prince2k3** [comment]
@tannernelson I used the tokenizer to be able to know what the sql statement looks like and format it they way I want once it got passed through. Also, to find the ? indexes for placing the values into them which could be used to rebuild the query.

**Comment by tannernelson** [comment]
Are the parameterized queries sent to the driver to be escaped by the support c lib, then?

**Comment by Prince2k3** [comment]
@tannernelson yeah the drivers (more so the db class) should handle the escaped strings. so for instance the sqlite will use bind to bind the values and use the parameterized query as the statement.

**Comment by tannernelson** [comment]
Ok, good! :)

**Pull request merged**